code/controllers/subsystem/move_loop.dm

```diff
@@ -0,0 +1,466 @@
+	var/atom/target
+
+/datum/move_loop/has_target/setup(atom/moving, delay, timeout, atom/chasing)
+	if(!..())
+		return FALSE
+	if(!isatom(chasing))
+		handle_delete()
+		return FALSE
+
+	target = chasing
+
+	if(!isturf(target))
+		RegisterSignal(target, COMSIG_PARENT_QDELETING, .proc/handle_no_target) //Don't do this for turfs, because of reasons
+
+	return TRUE
+
+/datum/move_loop/has_target/kill()
+	if(!isturf(target))
+		UnregisterSignal(target, COMSIG_PARENT_QDELETING)
+	return ..()
+
+/datum/move_loop/has_target/proc/handle_no_target()
+	SIGNAL_HANDLER
+	handle_delete()
+
+///Used for force-move loops
+/datum/move_loop/has_target/force_move
+
+/datum/move_loop/has_target/force_move/move()
+	moving.forceMove(get_step(moving, get_dir(moving, target)))
+
+///Base class of move_to and move_away, deals with the distance and target aspect of things
+/datum/move_loop/has_target/dist_bound
+	var/distance = 0
+
+/datum/move_loop/has_target/dist_bound/setup(atom/moving, delay, timeout, atom/chasing, dist = 0)
+	if(!..())
+		return FALSE
+	distance = dist
+
+	return TRUE
+
+
+/datum/move_loop/has_target/dist_bound/proc/check_dist()
+	return (get_dist(moving, target) >= distance) //If you get too close, stop moving closer
+
+/datum/move_loop/has_target/dist_bound/move()
+	if(!check_dist()) //If we're too close don't do the move
+		lasttick = round(lasttick - delay, 0.1) //Make sure to move as soon as possible
+		return FALSE
+	return TRUE
+
+///Wrapper around walk_to()
+/datum/move_loop/has_target/dist_bound/move_to
+
+/datum/move_loop/has_target/dist_bound/move_to/check_dist()
+	return (get_dist(moving, target) >= distance) //If you get too close, stop moving
+
+/datum/move_loop/has_target/dist_bound/move_to/move()
+	if(!..())
+		return
+	step_to(moving, target)
+
+///Wrapper around walk_away()
+/datum/move_loop/has_target/dist_bound/move_away
+
+/datum/move_loop/has_target/dist_bound/move_away/check_dist()
+	return (get_dist(moving, target) <= distance) //If you get too far out, stop moving away
+
+/datum/move_loop/has_target/dist_bound/move_away/move()
+	if(!..())
+		return
+	step_away(moving, target)
+
+///Used as a alternative to walk_towards
+/datum/move_loop/has_target/move_towards
+	///The turf we want to move into, used for course correction
+	var/turf/moving_towards
+	///Should we try and stay on the path, or is deviation alright
+	var/home = FALSE
+	///When this gets larger then 1 or smaller then -1 we move a turf
+	var/x_ticker = 0
+	var/y_ticker = 0
+	///The rate at which we move, between -1 and 1
+	var/x_rate = 1
+	var/y_rate = 1
+
+/datum/move_loop/has_target/move_towards/setup(atom/moving, delay, timeout, atom/chasing, home = FALSE)
+	if(!..())
+		return FALSE
+	src.home = home
+	update_slope()
+
+	if(home)
+		if(ismovable(target))
+			RegisterSignal(target, COMSIG_MOVABLE_MOVED, .proc/update_slope) //If it can move, update your slope when it does
+		RegisterSignal(moving, COMSIG_MOVABLE_MOVED, .proc/handle_move)
+
+	return TRUE
```

Comment:
```suggestion
	. = ..()
	if(!.)
		return FALSE
	src.home = home
	update_slope()

	if(home)
		if(ismovable(target))
			RegisterSignal(target, COMSIG_MOVABLE_MOVED, .proc/update_slope) //If it can move, update your slope when it does
		RegisterSignal(moving, COMSIG_MOVABLE_MOVED, .proc/handle_move)

```